table/tables/tables.go

```diff
@@ -637,9 +659,21 @@ func (t *TableCommon) AddRecord(sctx sessionctx.Context, r []types.Datum, opts .
 
 	for _, col := range t.WritableCols() {
 		var value types.Datum
-		// Update call `AddRecord` will already handle the write only column default value.
-		// Only insert should add default value for write only column.
-		if col.State != model.StatePublic && !opt.IsUpdate {
+		if col.ChangeStateInfo != nil && col.State != model.StatePublic {
+			// TODO: Check overflow or ignoreTruncate.
+			value, err = table.CastValue(sctx, r[col.DependencyColumnOffset], col.ColumnInfo, false, false)
+			if err != nil {
+				return nil, err
+			}
+			r = append(r, value)
+			row = append(row, value)
+			colIDs = append(colIDs, col.ID)
+			continue
+		}
```

Comment:
Maybe we can extract these lines into a function in the following PRs.

Reply:
OK~